gto/api.py

```diff
@@ -307,14 +308,14 @@ def find_versions_in_stage(
     )
 
 
-@clone_on_remote_repo
+@clone(repo_arg="repo", controller=is_url_of_remote_repo, controller_args=["repo"])
 def check_ref(repo: Union[str, Repo], ref: str):
     """Find out what have been registered/assigned in the provided ref"""
     reg = GitRegistry.from_repo(repo)
     return reg.check_ref(ref)
```

Comment:
I may be missing something here, but what is preventing us to do something like follows?

```suggestion
@clone(repo_arg="repo", controller=is_url_of_remote_repo, controller_args=["repo"])
def check_ref(repo: Union[str, Repo], ref: str):
    """Find out what have been registered/assigned in the provided ref"""
    with open_repo(repo) as r:
        reg = GitRegistry.from_repo(r)
        return reg.check_ref(ref)
```

Reply:
Don't you mean to remove decorator at all here?

Reply:
Sorry, fixed.

Reply:
We had a long discussion about it in https://github.com/iterative/gto/issues/25. We decided it's better to use decorator instead of writing this. It looked like a good way to do this at that moment. Maybe it's time to revisit that decision.

Reply:
Btw, if we need to also pass params like `--commit` and `--push` (to automatically commit and push commit or tag), could it work like `with open_repo(repo, commit=commit, push=push) as r: ...`?

Reply:
Ok, there are at least few ways to do it:
1. Leave `with open_repo(repo) as r:` for the user, as in DVC (`with dvc.api.open(...)`) https://dvc.org/doc/api-reference/open
```
with gto.api.open_repo(repo) as r:
    return gto.api.check_ref(...)
```
Cons: This will require Studio to rewrite API calls, and requires to add this `with ...` to all CLI commands. Pros: it allows to re-use cloned repo, and removes the repetitive `with gto.api.open_repo(repo) as r` part from API commands code.
2. Do it inside API calls - this is what @skshetry suggested above.
3. Do both. Provide `gto.api.open_repo` to do the 1st - to re-use cloned repo, and do as @skshetry suggested anyway - in case user supplies a remote repo.

Reply:
And I assume that pro for any of these approaches will be tests simplification, but haven't TAL at that yet.

Reply:
Also, this reminds me of this old request from Studio: when they read GTO repos, they have to iterate a lot on different commits. Bc they call GTO API, `GitRegistry` instance gets recreated for each API call, which slows down things a lot. If we could create not `gitpython.Repo` instance, but `GitRegistry`, that could solve the problem.

The problem with that was: `GitRegistry` provides a single entrypoint (of lower level than `gto.api`), but the state is actually kept in `BaseRegistryState` object. And `BaseRegistryState` instance is re-generated every time it's needed for `GitRegistry` to work with. E.g., some operations like `register` or `assign` can change the state, so sometimes `BaseRegistryState` inside `GitRegistry.register` or `GitRegistry.assign` calls. 🤯

Reply:
But probably, that is another issue that doesn't have to be coupled with reading remote repos. WDYT, @francesco086?

Reply:
btw, isn't `dvc.repo.Repo.open` in fact the same as `gto.registry.GitRegistry.from_repo`? They both return `cls(...)`, although `dvc.repo.Repo.open` clones repo if needed.
They're used to expose some abstraction of DVC repo / GTO repo: e.g. compare this usage https://github.com/iterative/dvc/blob/b2a275ee9e865dec6d632aa96ac90eb385fab468/dvc/api/data.py#L21 with this usage https://github.com/iterative/gto/blob/main/gto/api.py#L313 (missing context manager doesn't count).

We even had [a idea (though with a strange code snippet I must admit)](https://github.com/iterative/gto/issues/25#issuecomment-1246585717) to clone remote repo in `from_repo` method - and `dvc.repo.Repo.open` does the same, although delegates it to `dvc.external_repo` module.

Reply:
`dvc.external_repo` is just an internal implementation detail, and is legacy thing (we did not have scmrepo). Will be refactored when https://github.com/iterative/scmrepo/issues/41 is fixed.

Reply:
btw, @skshetry, have you maybe solved similar problems for Studio as I stated above? In our case the problem is that we need to read all Git tags with each API call - and this is time consuming. I understand you don't need to do that, you rather need to expose what's in a specific commit. But, maybe you still did some similar optimization? Looking for some examples, if they exist.

Reply:
One comment, if I understood correctly, actually the code should be:
```python
def check_ref(repo: Union[str, Repo], ref: str):
    """Find out what have been registered/assigned in the provided ref"""
    with open_repo(repo) as reg:
        return reg.check_ref(ref)
```
right?

Reply:
> If we could create not `gitpython.Repo` instance, but `GitRegistry`, that could solve the problem.

Very sorry, I am not able to follow: what is `GitRegistry`?

Reply:
Sorry for the late answer, yesterday was busy, I tried from the phone but it was a bad idea.

-----------------

Thanks for the explanations @aguschin , many things I was unaware of. From what I understood your suggestions lead to change so that in `cli` (not `api`!) we do
```python
with gto.registry.GitRegistry.from_repo(r) as reg:
    return reg.check_ref(ref)
```
and `check_ref` in `api` could be deleted...? which means that Studio developers will have to adjust their code-base in exchange for performance gain.
Is this what you are suggesting? With the following extra considerations:
1. `GitRegistry.from_repo` should support remote repos
2. `GitRegistry.from_repo` should allow the `commit` and `push` options (or, most likely, not, and find another mechanism)

The only warning sign that comes up to my mind, is that one should be careful not to break the single-responsibility principle, and don't mix all these functionalities together. The single responsibilities I see:
- read a gto registry from local
- clone remote repos
- commit changes
- push tags
- push commit
So, nothing against having a class that allows what I understood you were sketching, but let's be careful about keeping the functionalities distinct and re-usable :)

-----------------

> The problem with that was: GitRegistry provides a single entrypoint (of lower level than gto.api), but the state is actually kept in BaseRegistryState object. And BaseRegistryState instance is re-generated every time it's needed for GitRegistry to work with. E.g., some operations like register or assign can change the state, so sometimes BaseRegistryState inside GitRegistry.register or GitRegistry.assign calls. 

This sounds to me more of a general design problem?

-----------------

These are my two pennies, but I think you have a better overview.

Moving into something practical, so that perhaps I can amend what I have done, and leave general `gto` design decisions and changes to another time: I believe that the simple decorators in `git_utils` (forget the horrible names) respect the rule of doing one thing only, and can be combined with any other class or functions you may create in the future to enhance capabilities, while keeping concerns separated. Would it make sense if I finalize this module, and then use it in any way you consider more appropriate? What I mean is: `@clone_if_remote` could be used to decorate the `api.check_ref`, `GitRegistry.from_repo`, or even `cli.check_ref`. What do you think?

As a matter of fact I was even thinking of taking this `git_utils` out of `gto` and make it a stand-alone mini-library that one can re-use in many project to add support to remote repos.

Reply:
> From what I understood your suggestions lead to change so that in cli (not api!) we do
> ```
> with gto.registry.GitRegistry.from_repo(r) as reg:
>    return reg.check_ref(ref)
> ```
That would just what DVC does. We can remove API - for the sake of development simplicity. It basically just duplicates everything from GitRegistry, and adds only `show`, `history` - and those can be migrated to some "Presenter" class? Or maybe even to `BaseRegistryState` 🤔

Reply:
@francesco086, do you think it would make the tests much easier? I didn't have time to dig into your tests or try to rewrite them, so relying on your opinion here.

Reply:
I am sorry but I am lost, what would make tests easier? Option 1, 2, or 3 in https://github.com/iterative/gto/pull/296#discussion_r1016147255? Could you perhaps rephrase? I am not sure I understand what is it that you mean now.

Reply:
Specifically, what you suggested: use `with gto.registry.GitRegistry.from_repo(r) as reg:` instead of decorators.

Reply:
Ok, now I understand, thanks :)

The answer is: it depends on how we are going to add the remote functionalities to `GitRegistry.from_repo(...)`. For what concern the testing difficulties, this change means only to move the target (if we keep the decorators) :D

This would reduce the number of tests for the cloning functionality, because that would happen only in `from_repo()`. All other functionalities (commit, push) will have to be checked on all (or at least, most) of the `GitRegistry` methods. And if we keep using decorators, there will be the same difficulties.

But, to be honest, when I look at the code itself, I very much like the decorator approach, because there is a very neat separation of concern. What I find very cumbersome is to test them, but maybe it's me, I don't know how to do it in the right way. Perhaps this could be formulated in this way (in simple terms):
```python
@g
def f():
    ...
```
Suppose we are working TDD. Suppose that I fully tested `f` when `g` does nothing (e.g. in our case, if the repo is not remote) and I tested that when I decorate any function with `g`, everything works as expected. So, I can trust `f` and `g` taken singularly. What is the "right" way to ensure that `f` has been actually decorated with `g`? The tests I have before would pass whether `f` is decorated or not. What is the smallest test I need to add that would fail, and therefore would force me to decorate `f` with `g`?

So far the only answer to this question seemed to be `inspect.getsource` and check that there is a decorator. But I find this approach of parsing the source code a little weird...

One thing more (I hope not to confuse you, please just skip it if this is the case): I talked about TDD according to the London school. If we take the Chicago school, then probably we could say "we don't check the decorators with unittest because they are private methods, and rather check the functionalities in the public ones (`api`)." In practice this would mean to leave the tests of the `api` as they were, and delete the tests of the decorators in `git_utils`. Even though this may sound simple, it actually would lead to a higher complexity if one wants to test everything properly (one should check all possible combinations of inputs, and by adding `remote_or_not`, `commit`, and `push`, means 8 times more test). Much easier to test 2 cases (`remote_or_not`), then other 2 cases (`commit`), then other 2 cases (`push`), separately, and then make sure that they combine properly.

In short, if we change the decorators to context managers (notice: external to `api` and `registry`, so not by including them in `GitRegistry`), tests may get simpler. At the price of changing the code. If someone has a good idea to solve the testing problem I explained above with `f` and `g`, that would be ideal...
With this said, I have nothing against "moving the target" and decorate `GitRegistry` instead of the `api` functions, if this helps performance etc.

Reply:
My suggestion, considering all:
- move the remote repo functionalities into `GitRegistry` directly (so, suggestion from @skshetry will be realized)
- I am unsure whether to keep decorators or witch to context managers. When I am unsure, I prefer to be lazy, so eventually I think I would stick to the decorators.

But I leave the decision to you :)

What about grouping tests into classes, is that ok? Like in here: https://github.com/iterative/gto/pull/296/files#diff-dca8d5901c13c8fdb3912ed1fa77246f037bdda082b903fac553851476779773R31